### DISCUSSION TITLE
What is $K1 in form data with a server action?

### DISCUSSION THREAD

**Discussion by jess-into** [vercel/next.js#72809] [Help] [answered]
### Summary

When I call a server action I see an additional parameter in the request payload as follows:
`[\"$K1\",false]`

Does anyone know what this is used for or why it is added? And if it's possible to not include it. This is causing the azure frontdoor WAF to be triggered, the rule it is hitting is 'Detects basic SQL authentication bypass attempts 2/3'.

The match from the logs:
` [{"matchVariableName":"MultipartParamValue:0","matchVariableValue":"[\"$K1\",false]"}]`

**Comment by icyJoseph** [comment, marked as answer]
I think this comes from React code. In general, one has to remember that Server Actions are a React thing.

I found this in the React codebase, https://github.com/facebook/react/blob/3644f0bd214b7c08a522cc9212dfca902777bf8c/packages/react-client/src/ReactFlightReplyClient.js#L117-L120:

```js
function serializeFormDataReference(id: number): string {
  // Why K? F is "Function". D is "Date". What else?
  return '$K' + id.toString(16);
}
```

Hehe, it was funny. 

If you fire the server action by other means, such as an `onClick` on a button, then the `$K1` is not used.

**Reply by jess-into** [reply]
Thank you, that's really helpful!
